src/spi.rs

```diff
@@ -1639,32 +1745,104 @@ where
     }
 }
 
+fn spi_operations<'a>(
+    operations: impl Iterator<Item = Operation<'a>> + 'a,
+    chunk_size: usize,
+    duplex: Duplex,
+) -> impl Iterator<Item = SpiOperation> + 'a {
+    enum OperationsIter<R, W, T, I, D> {
+        Read(R),
+        Write(W),
+        Transfer(T),
+        TransferInPlace(I),
+        Delay(D),
+    }
+
+    impl<R, W, T, I, D> Iterator for OperationsIter<R, W, T, I, D>
+    where
+        R: Iterator<Item = SpiOperation>,
+        W: Iterator<Item = SpiOperation>,
+        T: Iterator<Item = SpiOperation>,
+        I: Iterator<Item = SpiOperation>,
+        D: Iterator<Item = SpiOperation>,
+    {
+        type Item = SpiOperation;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            match self {
+                Self::Read(iter) => iter.next(),
+                Self::Write(iter) => iter.next(),
+                Self::Transfer(iter) => iter.next(),
+                Self::TransferInPlace(iter) => iter.next(),
+                Self::Delay(iter) => iter.next(),
+            }
+        }
+    }
+
+    operations.flat_map(move |op| match op {
+        Operation::Read(words) => OperationsIter::Read(
+            spi_read_transactions(words, chunk_size, duplex, LineWidth::Single)
+                .map(SpiOperation::Transaction),
+        ),
+        Operation::ReadWithWidth(words, line_width) => OperationsIter::Read(
+            spi_read_transactions(words, chunk_size, duplex, line_width)
+                .map(SpiOperation::Transaction),
+        ),
+        Operation::Write(words) => OperationsIter::Write(
+            spi_write_transactions(words, chunk_size, LineWidth::Single)
+                .map(SpiOperation::Transaction),
+        ),
+        Operation::WriteWithWidth(words, line_width) => OperationsIter::Write(
+            spi_write_transactions(words, chunk_size, line_width).map(SpiOperation::Transaction),
+        ),
+        Operation::Transfer(read, write) => OperationsIter::Transfer(
+            spi_transfer_transactions(read, write, chunk_size, duplex)
+                .map(SpiOperation::Transaction),
+        ),
+        Operation::TransferInPlace(words) => OperationsIter::TransferInPlace(
+            spi_transfer_in_place_transactions(words, chunk_size).map(SpiOperation::Transaction),
+        ),
+        Operation::DelayNs(delay) => {
+            OperationsIter::Delay(core::iter::once(SpiOperation::Delay(delay)))
+        }
+    })
+}
+
 fn spi_read_transactions(
     words: &mut [u8],
     chunk_size: usize,
     duplex: Duplex,
+    line_width: LineWidth,
 ) -> impl Iterator<Item = spi_transaction_t> + '_ {
     words.chunks_mut(chunk_size).map(move |chunk| {
         spi_create_transaction(
             chunk.as_mut_ptr(),
             core::ptr::null(),
-            if duplex == Duplex::Half3Wire {
-                0
-            } else {
+            if duplex == Duplex::Full {
```

Comment:
Did you test this here? I think this is wrong from past memory. The underlying parameters here are total_length and rx_length. 

I think Half3Wire was the odd one here, but for HalfDuplex you have to set the total_length the same as the rx_length here. E.g the total length cannot be smaller than the rx_length.

Reply:
Yeah, I had to set this otherwise this assertion fails:

https://github.com/espressif/esp-idf/blob/46acfdce969f03c02b001fe4d24fa9e98f6adc5e/components/esp_driver_spi/src/gpspi/spi_master.c#L1067

Reply:
I did test it and it all works on my device (single, dual, quad). But honestly I’m pretty confused about how these fields are intended to be used, so it would be great to double check that.

Reply:
On what type of esp did you test it? By a quick glance it seams to only assert on non esp32/s2 variants. If that is the case and it works on esp32 differently we may need to add a cfg flag here for conditional compilation based on that.

Reply:
the corresponding flag would be `[#cfg(esp_idf_soc_spi_hd_both_inout_supported)]`

Reply:
Oh, this was on an ESP32-S3. Makes sense then— I won’t be able to test it properly though on an ESP32.

I’m surprised the API would work so differently though between the two devices though. The way it’s worded makes it sound like it’s an actual behavior difference, a hardware feature that either is or isn’t supported on different devices.

Reply:
Than for now lets add this cfg flag, and if its set set the length to the same as rx_length, and if its not present put a 0 in it like you did.

Reply:
If you look at the line above in esp-idf it also mentions this configuration not working on ESP32 when DMA is enabled. 

The wording “SPI half duplex mode does not support using DMA with both MOSI and MISO phases.”, to me, sounds like it’s a mode where you first do a write and then do a read (or vice versa). But nothing in this crate would map to that— you’d send separate Operations if you wanted that. So this behavior, to me, sounds like something that’s undesirable anyway.

Reply:
Yeah it would essentially mean this feature would only fully working on a s2 and its niche. Maybe there are some perf benefits in that specific scenario, though that's not relevant here.  

Let's bring the PR into master and see of other people start complaining shell we :D